### PR TITLE
rgw/rados: RGWRados::delete_objs_inline() uses AioThrottle

### PR DESCRIPTION
**Pull request by cbodley**
use concurrent deletes to speed up inline garbage collection

Fixes: https://tracker.ceph.com/issues/68134

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-api/81840/
> ERROR: test_list_enabled_module (tasks.mgr.dashboard.test_mgr_module.MgrModuleTest)

**Comment by cbodley** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_rados.cc [hunk 1]
```diff
@@ -5369,33 +5369,41 @@ std::tuple<int, std::optional<cls_rgw_obj_chain>> RGWRados::send_chain_to_gc(cls
...
+  auto obj = chain.objs.begin();
+
+  librados::IoCtx ioctx;
+  int ret = rgw_init_ioctx(dpp, get_rados_handle(), obj->pool, ioctx);
+  if (ret < 0) {
+    return;
+  }
+
+  // issue deletions in parallel, up to max_aio at a time
+  auto aio = rgw::make_throttle(cct->_conf->rgw_multi_obj_del_max_aio, y);
```
Comment (pritha-srivastava) [inline src/rgw/driver/rados/rgw_rados.cc:5387]: @cbodley : I have a basic question here - when I last worked on GC, the OSD was single threaded, and cls ops execute on an OSD, so how does parallel requests for these cls ops work here?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5387]: the rados cluster can only process one write op at a time per 'placement group'. each object name gets mapped to a specific pg, which is why we can expect cls operations on the same object to be atomic

but rados is highly scalable. each osd can hold up to `mon_max_pg_per_osd=250` pgs, and it's not uncommon for clusters to have hundreds of osds

gc is deleting many different tail objects here, so it's likely that most of them will map to different pgs which can be processed in parallel

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/143571/
> ninja: build stopped: interrupted by user.

https://jenkins.ceph.com/job/ceph-api/81848/
> FAIL: test_get_indiv_flag (tasks.mgr.dashboard.test_osd.OsdFlagsTest)

opened https://tracker.ceph.com/issues/68151

**Comment by cbodley** [comment]
jenkins test make check

File: src/rgw/driver/rados/rgw_rados.cc (hunk 1 above)
Comment (pritha-srivastava) [reply src/rgw/driver/rados/rgw_rados.cc:5387]: delete_objs_inline() takes care of tail objects belonging to a particular rgw object. So does each tail object map to a different rados object and hence get mapped to a different pg?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5387]: correct, a logical "s3 object" in rgw is composed of a head object and zero or more tail objects in rados

striping object data over 4MB rados objects allows GetObject and PutObject to stream these reads/writes in parallel

Comment (pritha-srivastava) [reply src/rgw/driver/rados/rgw_rados.cc:5387]: Thank you for the explanation.

**Review by pritha-srivastava — APPROVED** [review]

File: src/rgw/driver/rados/rgw_rados.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:5387]: thanks for review!

**Comment by cbodley** [comment]
```
src/test/cls_rgw/test_cls_rgw.cc:857:63: error: conversion from 'std::vector<cls_rgw_obj>::iterator' to non-scalar type 'std::__cxx11::list<cls_rgw_obj>::iterator' requested
  857 |     list<cls_rgw_obj>::iterator oiter = entry.chain.objs.begin();
      |                                         ~~~~~~~~~~~~~~~~~~~~~~^~
```

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-09-25_19:45:24-rgw-wip-cbodley-testing-distro-default-smithi/ with rerun https://pulpito.ceph.com/cbodley-2024-09-26_12:41:47-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/143969/
> The following tests FAILED:
	282 - jsonnet-bundler-build (Failed)
	292 - run-tox-jsonnet-check (Not Run)

**Comment by cbodley** [comment]
jenkins test make check

**Pull request merged**